src/NServiceBus.Core/Routing/EndpointInstance.cs

```diff
@@ -56,6 +56,24 @@ public EndpointInstance SetProperty(string key, string value)
         }
 
         /// <summary>
+        /// Creates a new endpoint instance with specified discriminator.
+        /// </summary>
+        /// <param name="discriminator">New discriminator value.</param>
+        public EndpointInstance SetDiscriminator(string discriminator)
```

Comment:
Both this and `SetEndpoint` really should be static factory methods since they return a new instance, and don't modify the existing one.

Looking at the rest of this class, `SetProperty` should be a factory method as well.